src/components/name-tag.js

```diff
@@ -0,0 +1,377 @@
+import anime from "animejs";
+import MovingAverage from "moving-average";
+import { getThemeColor } from "../utils/theme";
+import qsTruthy from "../utils/qs_truthy";
+import { findAncestorWithComponent } from "../utils/scene-graph";
+import { NAMETAG_VISIBILITY_DISTANCE_DEFAULT } from "../react-components/preferences-screen";
+import { THREE } from "aframe";
+import { setMatrixWorld } from "../utils/three-utils";
+
+const DEBUG = qsTruthy("debug");
+const NAMETAG_BACKGROUND_PADDING = 0.05;
+const NAMETAG_STATUS_BORDER_PADDING = 0.035;
+const NAMETAG_MIN_WIDTH = 0.6;
+const NAMETAG_HEIGHT = 0.25;
+const NAMETAG_OFFSET = 0.2;
+const TYPING_ANIM_SPEED = 150;
+const DISPLAY_NAME_LENGTH = 18;
+
+const ANIM_CONFIG = {
+  duration: 400,
+  easing: "easeOutQuart",
+  elasticity: 400,
+  loop: 0,
+  round: false
+};
+
+AFRAME.registerComponent("name-tag", {
+  schema: {},
+  init() {
+    this.store = window.APP.store;
+    this.displayName = null;
+    this.identityName = null;
+    this.isTalking = false;
+    this.isTyping = false;
+    this.isOwner = false;
+    this.isRecording = false;
+    this.isHandRaised = false;
+    this.volumeAvg = new MovingAverage(128);
+    this.isNametagVisible = true;
```

Comment:
I think this represents the desired state so maybe something like `shouldBeVisible` would be better. The actual state should live in `this.el.object3D.visibile`, and we should apply that once per frame, taking into consideration other overrides like hiding during model loading.

Reply:
Updated.